core/controllers/topics_and_skills_dashboard.py

```diff
@@ -117,6 +123,40 @@ def get(self) -> None:
         for topic_summary_dict in topic_summary_dicts:
             topic_summary_dict['classroom'] = topic_classroom_dict.get(
                 topic_summary_dict['id'], None)
+            topic = topic_fetchers.get_topic_by_id(topic_summary_dict['id'])
+            upcoming_chapters_count = 0
+            overdue_chapters_count = 0
+            total_chapters_counts = []
+            published_chapters_counts = []
+            for story_reference in topic.canonical_story_references:
+                story = story_fetchers.get_story_by_id(
```

Comment:
Definitely do not put a get_by_id() in a nested for loop, for the same reason as above.

See also fetch_multiple_entities_by_ids_and_models() which might be helpful if you can't do this in a simpler way.

Also, note that the bulk of this logic should be in the domain/services layer. Controllers are supposed to be thin and should not contain a lot of domain logic.

Reply:
Done.
Actually the logic is pretty small here (only 3 if conditions), its just due to the linting that the code takes more lines and seems complicated, but actually it is short, so I am not sure if we really need it to be in domain/services layer.

Reply:
By "logic", I also mean all the modifications of the dicts etc. So this still qualifies as "too much stuff for a controller", and you should have service functions that handle this.

Note also that when passing structured data from the domain layer to the controller layer, you should use a domain object with typed fields.

Reply:
Done.